Review of latest files.

jacobgardner opened the pull request:
Created a PR to a different branch because it's easier to review.

jrhii committed:
Began work

working on sockets and stuff.

jrhii committed:
Basic Monopoly Object

needs more work and more hammering out

jrhii committed:
added player class

not done

jacobgardner commented on src/main.js:
```diff
@@ -0,0 +1,32 @@
+var server = require('http').createServer();
```
Suggest changing these to not use require:

```
import  http from 'http';
import socketio from 'socket.io';

const server = http.createServer();
const io = socketio(server);
```

Just so it's consistent with everything else.

jacobgardner commented on src/main.js:
```diff
@@ -0,0 +1,32 @@
+
+io.on('connect', function(socket){
+  //assign user
+  socket.on('getAccountVal', function(playerId){
```
This approach is fine, but I'm still going to talk about the way I would've done it and why.  

Basically whenever a new turn starts, I would send the player the entire board state in one go.  So they would get a JSON string that looks like: 

```
{
    playerTurn: 'jim',
    gameState: {
        freeParkingCash: 7000,
    },
    playerState: {
        jim: {
            cashMoney: 582,
            cards: [3, 71, 12, 52], // basically some id corresponding with a card or somefin
            properties: [8, 12, 4],
            location: 42, // or 'jail', etc. 
            piece: 'topHat',
        },
        jake: {
            // Same stuff as above...
        },
    },
    propertyStates: [
        1: {
            isMortgaged: false,
            houses: 0, // 5 indicates hotel? 
            // whatever details
        },
        // etc. etc.
    ],
}
```

You could even put the full details for each property in propertyStates, and card details in cardStates.  That way you could even have different 'versions' of monopoly if you wanted with different sale prices/etc.  Or just so that the AIs can use it as a reference card.

So the reason I'd do this over requiring Remote Procedure Calls (RPC) like you're doing is that this is all the data AIs are going to want to request anyway.  So by just giving all of it to them at the start, you reduce the traffic to the server and I think it's a little simpler implementation.

jacobgardner commented on src/monopoly.js:
```diff
@@ -0,0 +1,48 @@
+  resetPlayers(){}
+
+  static rollDice(){
+    var dice = new Array(2);
```
Idiomatic JS dictates that we would probably use something like map for this.  As you use map more and more you'll become comfortable with it:

```
return (new Array(2)).map(() => {
    return Math.floor(Math.random() * 6) + 1;
});
```

jacobgardner commented on src/main.js:
```diff
@@ -0,0 +1,32 @@
+  });
+
+  socket.on('rollDice', function(){
+    dice = Monopoly.rollDice();//should also make monopoly member utilize roll action (e.g. move player on roll)
```
Always declare your variables with `let` or `const`.  If you don't your variables get declared globally and can lead to nasty bugs and memory inefficiency.  

The bugs occur when you name a variable that matches something already in the global object and overwrite it.  Memory wise your variables are typically marked for clean up at the end of their scope (I.e. the function they were made in. ). But if defined globally they're only cleaned up if you do it explicitly or when the program closes.

jacobgardner commented on src/player.js:
```diff
@@ -0,0 +1,26 @@
+    ToPlayer.setMoney(ToPlayer.getMoney() + amount);//give money
+  }
+
+  set setMoney(amount){//ask Jake about set and get
```
I can explain get and set, but for these, they're honestly not needed here.  

So typically in javascript you shouldn't have get or set functions.  Java/C++ require those because they don't support properties (the get/set keywords that I'll explain). 

So let's say we have a class that looks like this:

```
class Circle {
    constructor(radius) {
        this.radius = radius;
        this.area = Math.PI * radius * radius;
        this.circumference = 2 * Math.PI * radius;       
    }
}

const c = new Circle(10);
assert(c.radius == 10);
assert(c.area == 314.16);
assert(c.circumference == 6.2832);  // or whatever

// The above asserts should pass, but what if we want to change the radius:
c.radius = 100;
// Well the radius would update, but the area/circumference would still be the same. 
// We COULD turn area and circumference into functions, but we would end up losing 
//     the niceness that you get when they're properties.   Also extra functions pollute the code base.
//     This is what you do in Java because functions are the ONLY way to solve the problem.  But not so in 
//     Javascript

class Circle {
     constructor(radius) {
         this.radius = radius;
     }

     get area() {
         return this.radius * this.radius * Math.PI;
     }

     set area(newArea) {
          this.radius = Math.sqrt(newArea/Math.PI);
     }

     get circumference() {
          return this.radius * 2 * Math.PI;
     }

    set circumference(c) {
         this.radius = c/ (2 * Math.PI);
    }
}

// With this second version you can still access area/circumference like they were normal properties, but when you update them, radius will automatically update.   
```

TL;DR - Get rid of getMoney/setMoney and just use money.

jrhii replied:
so like this? : 

```
  static transferMoney(FromPlayer, ToPlayer, amount){
    if(amount > FromPlayer){
      //throw error, inssuficient funds;
    }

    FromPlayer.money = (FromPlayer.money - amount);//take money
    ToPlayer.money = (ToPlayer.money + amount);//give money
  }
```

to further clarify, no set/get functions if we don't need a function to make the result, and js doesn't support native private properties anyways?

jacobgardner replied:
Essentially.  I would suggest not making it a static method though and doing:

```
transferMoney(otherDude, amount) {
    if (amount > this.money) {  
// error
    } 

    this.money -= amount;
    otherDude.money += amount;

    return this; // Typically if your methods don't return anything, you return 'this' so you can chain methods together, if that makes sense:
    // Like you could do jim.transferMoney(jake, 500).transferMoney(bobby, 1);
}
```

jacobgardner commented on src/monopoly.js:
```diff
@@ -0,0 +1,48 @@
+  }
+
+  cleanBoard(){
+    PropertyList = new PropertyList;
```
You're not using `this` so you're setting these globally.

jacobgardner commented on src/monopoly.js:
```diff
@@ -0,0 +1,48 @@
+  loadPlayer(){}
+
+  runGame(){
+    cleanBoard();
```
I don't think these calls will work.  You need to do `this.cleanBoard();` etc.

jrhii committed:
post review corrections

not complete, may rewrite parts, so don't want to do anything
extraneous.

jrhii committed:
More Notes, replanning sockets

jrhii committed:
Reworking socket layout

planning for distrubiting boardstate every prompt.

jrhii committed:
Setting up test client

not sure how to use import in this case, also rollDice function not
returning numbers...should maybe just forget it for the time being, it
may not be final.

jrhii committed:
Something Like real progress

Need to work on assembling card and property lists next

jrhii committed:
minor change

added "return this" to (mostly) finished nonreturning functions

jrhii committed:
Properties File

begun integration of Monopoly Properties

jrhii committed:
minor stuff

It's kinda messy right now as I begin to mess around with events

jrhii committed:
turn function

updated turn iterations.

jrhii committed:
registy generator

tested a lot of code and made appropriate corrections.
added script to generate a registration IDs

jrhii committed:
minor changes

jacobgardner commented on src/dummyClient/dummyClient.js:
```diff
@@ -0,0 +1,20 @@
+import fs from 'fs';
+
+var socket = require('socket.io-client')('http://localhost:3000');
+socket.on('connect', function(){
```
Suggest using arrow function instead of function declaration.  Reasons below. 

```
socket.on('connect', function() {
   // Your code
});
```

Turns into:

```
socket.on('connect', () => {
    // Your code
});
```

Reasons: 
Function declarations (those that use the word function), when called create a new context, so if you want to access the `this` variable from the calling function you have to assign it to a new variable in the calling function and use that instead.   Arrow functions don't create a new context so you don't have to go through that rigamarole.

jacobgardner commented on src/dummyClient/dummyClient.js:
```diff
@@ -0,0 +1,20 @@
+socket.on('connect', function(){
+  console.log('connected');
+//  socket.emit('rollDice');
+  socket.on('sendBoardState', function(BoardState){
```
Use arrow function.

jacobgardner commented on src/dummyClient/dummyClient.js:
```diff
@@ -0,0 +1,20 @@
+
+  socket.emit('register', 'john', '9qArn');
+
+  socket.on('promptRoll', function(BoardState){
```
Use arrow function.

jacobgardner commented on src/dummyClient/trigger.js:
```diff
@@ -0,0 +1,6 @@
+var socket = require('socket.io-client')('http://localhost:3000');
+socket.on('connect', function(){
```
Use arrow function.

jacobgardner commented on src/dummyClient/dummyClient2.js:
```diff
@@ -0,0 +1,21 @@
+  console.log('connected');
+//  socket.emit('rollDice');
+
+  socket.on('sendBoardState', function(BoardState){
```
Use arrow function.

jacobgardner commented on src/dummyClient/dummyClient2.js:
```diff
@@ -0,0 +1,21 @@
+import fs from 'fs';
+
+var socket = require('socket.io-client')('http://localhost:3000');
+socket.on('connect', function(){
```
Use arrow function.

jacobgardner commented on src/dummyClient/dummyClient2.js:
```diff
@@ -0,0 +1,21 @@
+
+  socket.emit('register', 'Tim', 'q2p0O');
+
+  socket.on('promptRoll', function(BoardState){
```
Use arrow function.

jacobgardner commented on src/main.js:
```diff
@@ -1,29 +1,40 @@
-var server = require('http').createServer();
```
Use arrow functions throughout.

jacobgardner commented on src/main.js:
```diff
@@ -1,29 +1,40 @@
 
+const server = http.createServer();
+const io = socketio(server);
+const MonopolyGame = new Monopoly();
```
In general, instances of classes should start lowercase and classes should start uppercase.  That way it's easier to visually inspect code and know exactly what's a variable and what's a class.

jacobgardner commented on src/main2.js:
```diff
@@ -0,0 +1,16 @@
+import Property from "./property";
+import fs from 'fs';
+//import "./cardProps/properties.json";
+let PropertyArray = new Array(0);
```
These should probably be const.  You're not going to be reassigning them at any point.

const doesn't prevent you from changing an object's contents, just prevents you from assigning something new to a variable. 

Also start this offer lowercase.  Same reasons above.

jacobgardner commented on src/monopoly.js:
```diff
@@ -1,48 +1,158 @@
+
+export default class Monopoly {
   constructor(){
-    this.PropertyList = null;
```
These should all probably start lowercase.

jacobgardner commented on properties.json:
```diff
@@ -0,0 +1,644 @@
+    "houses": null,
+    "isMortgaged": null,
+    "isBuyable": null,
+    "functionID": 6
```
I'm only guessing here but functionID probably refers to a numeric ID which executes the logic for when someone lands on this space?  

If that's so, I'd go ahead an replace this with a string indicating the logic that should be executed.

type: "GO", type: "Property', type: "utility", etc. etc.

The extremely marginal speed increase you're getting by dealing with integers instead of strings isn't worth the confusion and burden of knowledge necessary to work with this JSON object.  

Jacob Gardner's Pillars of Development in order
- Write Self-Documenting Code - This is code where function names, object names, and variable names tell a clear story of what's going on without having to do reference lookups. 
- Write Clear Code - Spacing and shit such that function scope is clear to the developer, and other things that make code look uniform. 
- Write code that reduces burden of knowledge (1 and 2 help with this)
- Write fast code (supercedes top 3 during optimization stage and replacement is very well documented)

jacobgardner commented on src/stringGen.js:
```diff
@@ -0,0 +1,20 @@
+    let text = "";
+    const possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
+
+    for( let i=0; i < 5; i++ )
```
Most places that you work at will have coding standards that require any scope to have surrounding curly braces.   Because of that, I suggest doing it here.

jacobgardner commented on src/stringGen.js:
```diff
@@ -0,0 +1,20 @@
+    const possible = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
+
+    for( let i=0; i < 5; i++ )
+        text += possible.charAt(Math.floor(Math.random() * possible.length));
```
I find the standard random function a bitch to work with sometimes.  My preference would be to find a random library that lets me do something like:  `random.choose(possible, 5);`

jacobgardner commented on src/monopoly.js:
```diff
@@ -1,48 +1,158 @@
+    let highIndex = 0;
+
+    this.PlayerArray.forEach(function(thisPlayer, index){
+      let thisRoll = Monopoly.rollDice().reduce(( acc, cur ) => acc + cur, 0);
```
Nice.

Edit: I take that back.  Use const here.

jacobgardner commented on src/monopoly.js:
```diff
@@ -1,48 +1,158 @@
 
-    resetPlayers();
+  findFirstPlayer(){//TODO what if ties occur
+    let highRoll = 0;
```
Is it important to you that the random-player-chooser-majig simulates what players do in the game?  I.e. rolling dice to see who goes first then doing roll-offs. 

If not, I would just take the list of players, scramble them, and call that the order.  
https://www.npmjs.com/package/random-js

Has a shuffle method.  You can also use it to generate your random ids.  just a thought.  

It also has dice rolling shit.

jacobgardner commented on src/monopoly.js:
```diff
@@ -1,48 +1,158 @@
+//      thisGame.landOnFunction(ActivePlayer);TODO
+
+      if(diceArray.every(element => element == diceArray[0])){//probably too much.  allows for more than 2 dice
+        if(++ActivePlayer.doubles >= 3){
```
It took me like 3 times to read through this to understand what was going on.  I generally avoid the prefix incrementer/decrementer because it's hard to visually inspect and know immediately what's going on.  

Suggest changing to:

```
activePlayer.doubles += 1;
if (activePlayer.doubles >= 3) {
// stuff
}
```

jacobgardner commented on src/monopoly.js:
```diff
@@ -1,48 +1,158 @@
+          thisGame.runTurn(ActivePlayer);
+          return this;
+        }
+      }
```
I think if you make this: 

```
} else {
activePlayer.doubles = 0;
}
```

you can get rid of the other two return statements and just have them all flow to the one below.

jacobgardner commented on src/monopoly.js:
```diff
@@ -1,48 +1,158 @@
-  runTurn(){
-    
+  runTurn(ActivePlayer){
+    const thisGame = this;//because 'this' breaks with the event
```
If you use an arrow function below, you wont need to use thisGame.

jacobgardner commented on src/monopoly.js:
```diff
@@ -1,48 +1,158 @@
+        if(++ActivePlayer.doubles >= 3){
+          ActivePlayer.doubles = 0;
+          thisGame.goToJail(ActivePlayer);
+          return this;
```
Just because this is a really good example of where arrow functions are awesome, I'm going to comment on it again.  
`this` refers to the wrong context currently so when you `return this`, it's not the `this` instance of `Monopoly`, it's the `this` created by the socketio callback which I'm not sure is anything.  If you use arrow functions, though,  `this` will return to the `Monopoly` instance you're using like you're hoping.

jacobgardner commented on src/monopoly.js:
```diff
@@ -1,48 +1,158 @@
+      if(regArray.length == this.PlayerArray.length){
+        console.log('runGame');
+        this.runGame();
+        return this;
```
You can get rid of these two `return this;` since they'll both fall down into that last one.

jacobgardner commented on src/player.js:
```diff
@@ -1,26 +1,23 @@
-    if(amount > FromPlayer){
-      //throw error, inssuficient funds;
+  movePlayer(diceArray, boardLength){//DONE
+    this.position += diceArray.reduce(( acc, cur ) => acc + cur, 0);
```
You actually do this pattern a few times.  If this continues you might consider building a new `Dice` object that has access to the individual rolls in addition to a nice `.sum()` function.

jrhii committed:
Revert "minor changes"

This reverts commit c91e381239e0a50b764b389413955a5bb0c46d1d.

jrhii committed:
Revert "Revert "minor changes""

This reverts commit 3901fb754b32ad25db48d841642c7e9638173d8f.

jrhii committed:
code review corrections

made corrections pre review.  fixed emitter leak.

jrhii committed:
minor fix

i wrote a booboo

jrhii committed:
working on land on functions

very incomplete

jrhii committed:
main.js commit

same as last commit, forgot to include main

jacobgardner commented on src/landOnFunctions.js:
```diff
@@ -0,0 +1,57 @@
+  if(this.ownerID == null){
+    emitter.emit(('promptBuy' + activePlayer.socketID), this.position);
+  }
+  esle{
```
ESLE

jacobgardner commented on src/monopoly.js:
```diff
@@ -134,6 +140,41 @@ export default class Monopoly {
+    this.propertyArray = JSON.parse(fs.readFileSync('properties.JSON'));
+    this.propertyArray.forEach((property) => {
+      switch (property.functionID) {
+        case 1:
```
Still suggest replacing functionID with string that represents the property

Then your case would look like:

```
case 'standard': 
    property.landOnFunction = standardProperty; 
    break;
case 'railroad': 
    property.landOnFunction = railroad;
    break;
// etc. etc.
```

Just makes the code clearer.

jacobgardner replied:
Also, an alternative to the switch statement is using  map pointing to the functions:

```
const PROPERTY_FUNCTIONS = {
    'standard': standardProperty,
    'railroad': railroad,
    'utility': utility,
    // etc. 
};
```

Then in here:

```
property.landOnFunction = PROPERTY_FUNCTIONS[property.type];
```

This is more efficient than a switch statement.

jrhii replied:
Yes, My plan was to replace the ID numbers with strings once I finished testing some of the function concepts themselves.  Will look into the alternative solution now

jacobgardner commented on src/player.js:
```diff
@@ -9,6 +9,7 @@ export default class Player{
     this.jailFreeCards = null; //change to list or something;
     this.isJailed = false;
     this.doubles = null;//
```
Does it make more sense to initialize all these with like 0?

jacobgardner commented on src/monopoly.js:
```diff
@@ -64,39 +64,45 @@ export default class Monopoly {
     this.toggleListenersOn();
     console.log('Toggling Listers\n promptRolls for ' + activePlayer.person);//test
-    this.emitter.emit('promptRoll', activePlayer, () => {});
+    this.emitter.emit(('promptRoll' + activePlayer.socketID) , activePlayer, () => {});
```
I like string interpolation in es6 opposed to manually adding together strings:

```
`promptRoll${activePlayer.socketID}`
```

String interpolation only occurs when you use backticks ```.

jacobgardner commented on src/monopoly.js:
```diff
@@ -64,39 +64,45 @@ export default class Monopoly {
     });
 
     this.toggleListenersOn();
     console.log('Toggling Listers\n promptRolls for ' + activePlayer.person);//test
```
Same here as my comment below:

```
console.log(`Toggling Listeners\n promptRolls for ${activePlayer.person}.`);
```

jrhii committed:
function functionality added; documentation begun

filled out payment and buying prompts, began client doc

jrhii committed:
updated landOnFunctions

and revamping random
